README.md
# Introduction

This tool allows for loading of environment variables similar to 
`symfony/dotenv`, only then for Docker Compose environments. Loading 
environment variable this way should help with a more consistent development
process.

# Installation

Checkout the project and add `dev` to `$PATH`.

# Usage

Ensure `docker-compose.yml`, `.env` and `.env.dev` are available and configured. For instance, you could set the right `PHP_VERSION` for the application that you are running. Also, set the `DNS_NAME` to your preference. Note that `<SUBDOMAIN>.dev.mediacthq.nl` is not a valid domain, so you need to change this in any case. You could use `example.dev.mediacthq.nl` (which automatically resolves to 127.0.0.1). But you could also use `magento.local`, in which case you will need to make sure yourself this hostname is resolvable. For instance, by adding a line with `127.0.0.1 magento.local` to your `/etc/hosts` file. 

Then start the environment with `dev up`. Confirm that everything has started with `dev ps`.

It's possible to install template environments from this project using `dev init`, which will install useful files for a quick setup. If you don't have a `docker-compose.yml` file yet, this is the recommended way.

# Updating Templates

Whenever you update a file for a certain template like magento-2.
You should also update the `version` of the `docker-compose.yml`. You can find it here:
```yaml
x-custom:
  version: 1.x.x
  type: magento-2
```
Whenever you update a template in a project make sure you revert the custom changes.

You know when you have to update the templates when you see the message:
```yaml
The current docker template is outdated, 
please run "dev init TEMPLATE_NAME -f" to update it.
```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.4.4]
- Fix issue with AMQP >= 3.9.0 refusing to start with deprecated ENV variable names

## [1.4.3]
### Fixed
- Issue resolve issue where newest version after 8.0.28-19.
- Change image of nginx-fpm to one where xdebug works out of the box.

## [1.4.2]
### Fixed
- Issue where the nginx image was wrongly configured

## [1.4.1]
### Security
- [INTAINTD-338] Additionally fixed the port bindings for AMQP in the magento-2 template

## [1.4.0]
### Security
- [INTAINTD-338] Forced binding of all exposed ports to localhost (IPv4). Note that if you have IPv6 enabled and using localhost you might need to explicitly use 127.0.0.1

## [1.3.2]
### Fixed
- [INTAINTD-259] Update `magento-2-console` images to latest (supported) version.

## [1.3.1]
### Added
- M2: Rabbitmq now has its own `volume` and hostname so queues persist between a `dev down && dev up`.

## [1.3.0]
### Added
- Extended PHP environment for generic use that is more inline with the current magento stack.

## [1.2.0]
### Changed
- M2: Redis now has its own `volume` so config cache and sessions persist between a `dev down && dev up`.

## [1.1.0]
### Changed
- Instead of adding Rabbitmq-server this will also add a UI that can be reached using 
localhost:15672 in your web browser.

## [1.0.1]
### Changed
- Template for M2 now also enables xdebug of composer.
- Updated the version of the template for M2 template.
- Changed mysql from 5.6 to 5.7 for M1 template.
- Changed php from 7.2 to 7.3.

### Removed
- Unused variable.


traefik/scripts/makehosts.py
import json, urllib2

# Retrieve all load balancers
response = urllib2.urlopen('http://127.0.0.1:8080/api/tcp/services')
tcpServices = json.loads(response.read())

for service in tcpServices:
    include = False
    for routerName in service['usedBy']:
        # Fetch the routers so the DNS entries can be generated
        routerResponse = urllib2.urlopen('http://127.0.0.1:8080/api/tcp/routers/' + routerName)
        tcpRouter = json.loads(routerResponse.read())
        # Check if the router uses the `mysql` entry point
        if "mysql" in tcpRouter['using']:
            # Tidy up the HostSNI rule.
            rule = tcpRouter['rule'].replace('HostSNI(`', '').replace('`)', '').replace('`, `', '`,`')
            for dnsEntry in rule.split('`,`'):
                # This is one of the required entries for HostSNI so non TLS TCP connections are send here.
                if dnsEntry != '*':
                    # Generate and print the entry for the hosts file.
                    print service['loadBalancer']['servers'][0]['address'].replace(':3306', '') + '    ' + dnsEntry


templates/magento-2/app/etc/env.php
<?php
return [
    'backend' =>
        [
            'frontName' => 'admin',
        ],
    'crypt' =>
        [
            'key' => 'development',
        ],
    'session' => [
        'save' => 'redis',
        'redis' => [
            'host'                  => 'redismaster',
            'port'                  => '6379',
            'password'              => '',
            'timeout'               => '2.5',
            'persistent_identifier' => '',
            'database'              => '2',
            'compression_threshold' => '2048',
            'compression_library'   => 'gzip',
            'log_level'             => '1',
            'max_concurrency'       => '6',
            'break_after_frontend'  => '5',
            'break_after_adminhtml' => '30',
            'first_lifetime'        => '600',
            'bot_first_lifetime'    => '60',
            'bot_lifetime'          => '7200',
            'disable_locking'       => '0',
            'min_lifetime'          => '60',
            'max_lifetime'          => '2592000',
        ]
    ],
    'db' =>
        [
            'table_prefix' => '',
            'connection' =>
                [
                    'default' =>
                        [
                            'host' => '{{database_host}}',
                            'dbname' => '{{database_name}}',
                            'username' => '{{database_user}}',
                            'password' => '{{database_password}}',
                            'model' => 'mysql4',
                            'engine' => 'innodb',
                            'initStatements' => 'SET NAMES utf8;',
                            'active' => '1',
                        ],
                ],
        ],
    'resource' =>
        [
            'default_setup' =>
                [
                    'connection' => 'default',
                ],
        ],
    'x-frame-options' => 'SAMEORIGIN',
    'MAGE_MODE' => 'production',
    'cache_types' =>
        [
            'config' => 1,
            'layout' => 1,
            'block_html' => 1,
            'collections' => 1,
            'reflection' => 1,
            'db_ddl' => 1,
            'eav' => 1,
            'customer_notification' => 1,
            'full_page' => 1,
            'config_integration' => 1,
            'target_rule' => 1,
            'config_integration_api' => 1,
            'translate' => 1,
            'config_webservice' => 1,
        ],
    'install' =>
        [
            'date' => '{{rfc2822_date}}',
        ],
    'system' => [
        'default' => [
            'web' => [
                'unsecure' => [
                    'base_url' => '{{base_url}}',
                    'base_link_url' => '{{base_url}}',
                    'base_static_url' => '{{base_url}}static/',
                    'base_media_url' => '{{base_url}}media/',
                ],
                'secure' => [
                    'base_url' => '{{base_url}}',
                    'base_link_url' => '{{base_url}}',
                    'base_static_url' => '{{base_url}}static/',
                    'base_media_url' => '{{base_url}}media/',
                ],
                'seo' => [
                    'use_rewrites' => '1',
                ]
            ],
            'admin' => [
                'security' => [
                    'use_form_key' => '0',
                    'admin_account_sharing' => '1',
                ],
            ],
            'cms' => [
                'wysiwyg' => [
                    'enabled' => '0',
                ],
            ],
            'system' => [
                'full_page_cache' => [
                    'caching_application' => '{{caching_application}}',
                    'varnish' => [
                        'backend_host' => '127.0.0.1',
                        'backend_port' => '8080'
                    ]
                ]
            ]
        ]
    ],
    'cache' => [
        'frontend' =>[
            'default' =>[
                'backend'         => 'Cm_Cache_Backend_Redis',
                'backend_options' =>[
                    'server'        => '127.0.0.1',
                    'port'          => '6379',
                    'database'      => '0',
                ],
            ],
            'page_cache' =>[
                'backend'         => 'Cm_Cache_Backend_Redis',
                'backend_options' =>[
                    'server'        => '127.0.0.1',
                    'port'          => '6379',
                    'database'      => '1',
                    'compress_data' => '0',
                ],
            ],
        ],
    ],
    'http_cache_hosts' => [
        [
            'host' => '127.0.0.1',
            'port' => '6081'
        ]
    ]
];


templates/magento-2/env-dev.php
<?php

return [
    'backend' => [
        'frontName' => getenv('BACKEND__FRONT_NAME'),
    ],
    'crypt' => [
        'key' => getenv('CRYPT__KEY'),
    ],
    'db' => [
        'connection' => [
            'indexer' => [
                'host' => getenv('DB__CONNECTION__INDEXER__HOST'),
                'dbname' => getenv('DB__CONNECTION__INDEXER__DBNAME'),
                'username' => getenv('DB__CONNECTION__INDEXER__USERNAME'),
                'password' => getenv('DB__CONNECTION__INDEXER__PASSWORD'),
                'model' => 'mysql4',
                'engine' => 'innodb',
                'initStatements' => 'SET NAMES utf8;',
                'active' => '1',
                'persistent' => null,
            ],
            'default' => [
                'host' => getenv('DB__CONNECTION__DEFAULT__HOST'),
                'dbname' => getenv('DB__CONNECTION__DEFAULT__DBNAME'),
                'username' => getenv('DB__CONNECTION__DEFAULT__USERNAME'),
                'password' => getenv('DB__CONNECTION__DEFAULT__PASSWORD'),
                'model' => 'mysql4',
                'engine' => 'innodb',
                'initStatements' => 'SET NAMES utf8;',
                'active' => '1',
            ],
        ],
        'table_prefix' => ''
    ],
    'resource' => [
        'default_setup' => [
            'connection' => 'default',
        ],
    ],
    'session' => [
        'save' => 'redis',
        'redis' => [
            'host' => getenv('SESSION__REDIS__HOST'),
            'port' => getenv('SESSION__REDIS__PORT'),
            'password' => getenv('SESSION__REDIS__PASSWORD'),
            'timeout' => '2.5',
            'persistent_identifier' => '',
            'database' => getenv('SESSION__REDIS__DATABASE'),
            'compression_threshold' => '2048',
            'compression_library' => 'gzip',
            'log_level' => '1',
            'max_concurrency' => '6',
            'break_after_frontend' => '5',
            'break_after_adminhtml' => '30',
            'first_lifetime' => '600',
            'bot_first_lifetime' => '600',
            'bot_lifetime' => '60',
            'disable_locking' => '0',
            'min_lifetime' => '60',
            'max_lifetime' => '2592000',
            'sentinel_master' => '',
            'sentinel_servers' => '',
            'sentinel_connect_retries' => '5',
            'sentinel_verify_master' => '0',
        ]
    ],
    'x-frame-options' => 'SAMEORIGIN',
    'cache_types' => [
        'config' => 1,
        'layout' => 1,
        'block_html' => 1,
        'collections' => 1,
        'reflection' => 1,
        'db_ddl' => 1,
        'compiled_config' => 1,
        'eav' => 1,
        'customer_notification' => 1,
        'config_integration' => 1,
        'config_integration_api' => 1,
        'target_rule' => 1,
        'google_product' => 1,
        'full_page' => 1,
        'config_webservice' => 1,
        'translate' => 1,
        'vertex' => 1,
    ],
    'install' => [
        'date' => '$(date -R)',
    ],
    'queue' => [
        'amqp' => [
            'host' => getenv('QUEUE__AMQP__HOST'),
            'port' => getenv('QUEUE__AMQP__PORT'),
            'user' => getenv('QUEUE__AMQP__USER'),
            'password' => getenv('QUEUE__AMQP__PASSWORD'),
            'virtualhost' => getenv('QUEUE__AMQP__VIRTUALHOST'),
        ],
    ],
    'http_cache_hosts' => [
        0 => [
            'host' => getenv('HTTPS_CACHE_HOSTS__0__HOST'),
            'port' => getenv('HTTPS_CACHE_HOSTS__0__PORT'),
        ],
    ],
    'cache' => [
        'frontend' => [
            'default' => [
                'id_prefix' => '40d_',
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => getenv('CACHE__FRONTEND__DEFAULT__BACKEND_OPTIONS__HOST'),
                    'port' => getenv('CACHE__FRONTEND__DEFAULT__BACKEND_OPTIONS__PORT'),
                    'database' => getenv('CACHE__FRONTEND__DEFAULT__BACKEND_OPTIONS__DATABASE'),
                    'password' => getenv('CACHE__FRONTEND__DEFAULT__BACKEND_OPTIONS__PASSWORD'),
                    'compress_data' => '1',
                    'compression_lib' => '',
                ],
            ],
            'page_cache' => [
                'id_prefix' => '40d_',
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => getenv('CACHE__FRONTEND__PAGE_CACHE__BACKEND_OPTIONS__HOST'),
                    'port' => getenv('CACHE__FRONTEND__PAGE_CACHE__BACKEND_OPTIONS__PORT'),
                    'database' => getenv('CACHE__FRONTEND__PAGE_CACHE__BACKEND_OPTIONS__DATABASE'),
                    'password' => getenv('CACHE__FRONTEND__PAGE_CACHE__BACKEND_OPTIONS__PASSWORD'),
                    'compress_data' => '0',
                    'compression_lib' => '',
                ],
            ],
        ],
    ],
    'system' => [
        'default' => [
            'twofactorauth' => [
                'general' => [
                    'disabled_users' => 'mediact*'
                ]
            ]
        ]
    ]
];
